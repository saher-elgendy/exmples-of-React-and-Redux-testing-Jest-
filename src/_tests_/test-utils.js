import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares);


//creating a custom render function so we can wrap our App with
//react-redux provider
const render = (ui, initialState) => {
  const store = mockStore(initialState);
  //App wrapper and mocked store passed to it
  const Wrapper = ({ children }) => {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }