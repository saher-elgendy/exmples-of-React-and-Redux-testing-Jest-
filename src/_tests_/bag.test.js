import React from 'react';
import Bag from '../components/bag';
import '@testing-library/jest-dom/extend-expect'
import { render } from './test-utils'
import { fireEvent, screen } from '@testing-library/react'

describe('bag', () => {
    // const {getByTestId} = screen
    let state = {
        bag: [],
    }


    it('render correctly when the use click the bag icon', () => {
        const { container } = render(<Bag />, state);
        expect(container.firstChild).toMatchSnapshot();

        fireEvent.click(screen.getByTestId(/bag-image/));
        expect(screen.getByTestId(/bag-content/)).toBeInTheDocument()
    });

    it('show a message if no books found in the bag', () => {
        render(<Bag />, state);
        fireEvent.click(screen.getByTestId(/bag-image/));
        expect(screen.queryByText(/No books/)).toBeInTheDocument()
    })

    it('show a list of books if books exist', () => {
        state = {
            bag: [
                {
                    name: 'Power of your subconscious mind',
                    author: 'Joseph Murphy',
                    cover: 'https://images-na.ssl-images-amazon.com/images/I/81MArWaiw1L.jpg'
                },
                {
                    name: 'The Secret',
                    author: 'Ronda Bayrne'
                }
            ]
        };

        render(<Bag />, state);
        fireEvent.click(screen.getByTestId(/bag-image/));
        //message 'No books found' shouldn't exist any more
        expect(screen.queryByText(/No books/)).not.toBeInTheDocument();
        expect(screen.getByText(/Power/)).toBeInTheDocument();
        expect(screen.getByText(/Secret/)).toBeInTheDocument();
        // expect(screen.getByAltText(/book_cover/)).toBeInTheDocument()
    })

    it('shows the correct books count in the bag', () => {
        //render bag with two books, the count should be 2
        state = {
            bag: [
                {
                    name: 'Power of your subconscious mind',
                    author: 'Joseph Murphy'
                },
                {
                    name: 'The Secret',
                    author: 'Ronda Bayrne'
                }
            ]
        }
        render(<Bag />, state);
        expect(screen.getByTestId(/books-count/)).toHaveTextContent('2');
    })
});