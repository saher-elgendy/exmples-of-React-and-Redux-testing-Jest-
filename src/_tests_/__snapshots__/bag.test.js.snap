// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bag render correctly when the use click the bag icon 1`] = `
<div
  class="bag"
>
  <div
    class="bag-container"
  >
    <span
      class="books-count"
      data-testid="books-count"
    >
      0
    </span>
    <img
      alt="bag"
      data-testid="bag-image"
      src="icon.png"
    />
  </div>
</div>
`;
